### PR TITLE
update HACKING.pod

### PR DESCRIPTION
**Pull request by cgzones**
This is just a first draft. Please discuss.
Related: #1014, #1051, #1060, #1061

### CODE REVIEW

File: HACKING.pod [hunk 1]
```diff
@@ -158,9 +188,31 @@ for more on POD.
...
 =head2 WRITING
 
 The POD should be defined all in one place. For plugins you place it at
 the top, else at the bottom.
 
+For scripts the recommend headers are (in this order):
+
+  NAME, SYNOPSIS, DESCRIPTION, REQUIRED ARGUMENTS, OPTIONS, EXIT STATUS,
+  CONFIGURATION, FILES, VERSION, BUGS AND LIMITATIONS, AUTHOR,
+  LICENSE AND COPYRIGHT
```
Comment (cgzones) [inline HACKING.pod]: those are just suggestions, subject to change

File: HACKING.pod [hunk 2]
```diff
@@ -118,37 +121,64 @@ helps. ASCII art is a must ;-)
...
+style and a tool-based enforcement. The exact configuration can be seen in the
+file I<.perltidyrc>. Please try to use it before any contribution by running
+C<make apply-formatting>.
+
+
+=head2 PERLCRITIC
+
+B<perlcritic> is used to comply to some wide-spread recommendations. The exact
+configuration can be seen in the file I<.perlcriticrc>. Please try to comply
+with those rules by running c<make lint>.
```
Comment (cgzones) [inline HACKING.pod]: see #1051 for questions about perlcritic severity 4 warnings

File: HACKING.pod [hunk 3]
```diff
@@ -158,9 +188,31 @@ for more on POD.
...
+
+  NAME, SYNOPSIS, DESCRIPTION, REQUIRED ARGUMENTS, OPTIONS, EXIT STATUS,
+  CONFIGURATION, FILES, VERSION, BUGS AND LIMITATIONS, AUTHOR,
+  LICENSE AND COPYRIGHT
+
+For modules the recommend headers are (in this order):
+
+  NAME, SYNOPSIS, DESCRIPTION, SUBROUTINES/METHODS,
+  CONFIGURATION AND ENVIRONMENT, DEPENDENCIES, BUGS AND LIMITATIONS, AUTHOR,
+  LICENSE AND COPYRIGHT
```
Comment (cgzones) [inline HACKING.pod]: those are just suggestions, subject to change

File: HACKING.pod [hunk 4]
```diff
@@ -158,9 +188,31 @@ for more on POD.
...
+
+For plugins the recommend headers are (in this order):
+
+  NAME, APPLICABLE SYSTEMS, CONFIGURATION, USAGE, MAGIC MARKERS, BUGS, AUTHOR,
+  LICENSE AND COPYRIGHT
```
Comment (cgzones) [inline HACKING.pod]: those are just suggestions, subject to change

File: HACKING.pod [hunk 5]
```diff
@@ -118,37 +121,64 @@ helps. ASCII art is a must ;-)
...
- };
- if ($@) {
-     # Handle exception
- }
+=head2 PERLTIDY
 
+We use B<perltidy> to automatically format perl code. This ensure a consistent
+style and a tool-based enforcement. The exact configuration can be seen in the
+file I<.perltidyrc>. Please try to use it before any contribution by running
+C<make apply-formatting>.
```
Comment (cgzones) [inline HACKING.pod]: see #1014

**Comment by sumpfralle** [comment]
Great work!

I think, you can omit the link to https://munin-monitoring.org/wiki - the guide should be sufficient.
Regarding the usual documentation headers I see an overlap between `AUTHOR` and `LICENSE AND COPYRIGHT`, but I have no strong opinions about that.
Thank you!

File: HACKING.pod [hunk 6]
```diff
@@ -22,13 +22,39 @@ here. However, all new code should be made to comply.
...
 =head1 GETTING STARTED
 
-In the B<dev_scripts> directory you will find scripts that are meant to be
-useful for developing. Most of them are tools for creating and using
-Munin in a sandbox.
+
+=head2 GITHUB
+
+The main development of Munin is happening on L<https://github.com>. With a
+free account you can report feature requests and issues
```
Comment (sumpfralle) [inline HACKING.pod]: Nitpicking: I would not like to the see the word `free` attached to a proprietary service.

How about "After registering an account"?

File: HACKING.pod [hunk 7]
```diff
@@ -22,13 +22,39 @@ here. However, all new code should be made to comply.
...
+pull requests (L<https://github.com/munin-monitoring/munin/pulls>).
+
+There are two interesting branches:
+
+=over
+
+=item I<stable-2.0>
+
+As the name suggests this is the stable branch for versions 2.0.x.
+Please base your plugin and bug-fix patches on this branch.
```
Comment (sumpfralle) [inline HACKING.pod]: Indeed I usually advertised this policy (plugin improvements go to `stable-2.0`). But @steveschnepp and me just realized, that we have diverging opinions on that matter. I hope, we will find the time to discuss this tomorrow in the weekly meeting. Thus the final text of this paragraph will need to be decided in IRC and not in this pull request, I guess. (just my opinion - we will see)

I always had a bad feeling about telling people to do some rebase without being able to point to a written documentation detailing this. Thank you for reducing this pain in the future!

File: HACKING.pod [hunk 8]
```diff
@@ -12,99 +12,132 @@ This is the guide for Munin internals contributors (developers,
...
-editor magic to ease the pain.  That said, nothing should be vertical-aligned.
-Only indentated correctly.
+Perl code should be written in tainted-mode with the settings I<use strict;>
+and I<use warnings;>. There should be no perl warnings when running Munin.
 
-When a subblock is started, always use one level.
+=item C<Spaces vs Tabs>
 
-=item C<Braces>
+We prefer indentation via spaces. Tab-based indentation is usually relying on
```
[the remark this answers is not in the document]
Comment (cgzones) [reply HACKING.pod]: no? but most of the code is using spaces

File: HACKING.pod (hunk 7 above)
Comment (sumpfralle) [reply HACKING.pod]: Update of our discussion: during the [IRC meeting on the 5th of September](http://meetbot.debian.net/munin/2018/munin.2018-09-05-18.29.log.html) we agreed on only applying fixes (both for core and for plugins) to `stable-2.0` from now own. These changes will be merged into `master` from time to time. All other changes go directly to `master`.

File: HACKING.pod (hunk 8 above)
[the remark this answers is not in the document]
Comment (steveschnepp) [reply HACKING.pod]: Well, you have a point : it's about 65% of spaces and 35% of tabs. 

```
~/src/munin$ grep -R '^ ' script/ lib/ plugins/ | wc -l                                                                                                                                                             
26066
~/src/munin$ grep -R '^        ' script/ lib/ plugins/ | wc -l                                                                                                                                                      
14301
```

Yet most of the *new* code is using tabs, as I wrote it. But I agree, the codebase is huge, and the ratio was very similar in 2.0. (70% spaces & 30% tabs)

**Comment by sumpfralle** [comment]
I think, there are two open issues:
1. @cgzones: would you please phrase the usage of the two branches following our IRC discussion?
2. @steveschnepp: how about the tab/space issue? I think, we all can live with every approach. Even something like *follow the overall style of the file in question* would be sufficient. But we should write *something* about it :)

**Comment by cgzones** [comment]
updated the branch and tab/space section

there are still some TODO's left...

**Comment by steveschnepp** [comment]
Let's merge it

**Pull request merged**